radixdlt-engine/src/main/java/com/radixdlt/application/system/construction/EpochConstructionState.java

```diff
@@ -0,0 +1,334 @@
+
+import com.google.common.collect.Streams;
+import com.google.common.primitives.UnsignedBytes;
+import com.radixdlt.application.system.scrypt.ValidatorScratchPad;
+import com.radixdlt.application.system.state.EpochData;
+import com.radixdlt.application.system.state.RoundData;
+import com.radixdlt.application.system.state.ValidatorBFTData;
+import com.radixdlt.application.system.state.ValidatorStakeData;
+import com.radixdlt.application.tokens.DelegatedResourceInBucket;
+import com.radixdlt.application.tokens.state.ExitingStake;
+import com.radixdlt.application.tokens.state.PreparedStake;
+import com.radixdlt.application.tokens.state.PreparedUnstakeOwnership;
+import com.radixdlt.application.validators.state.ValidatorData;
+import com.radixdlt.application.validators.state.ValidatorFeeCopy;
+import com.radixdlt.application.validators.state.ValidatorOwnerCopy;
+import com.radixdlt.application.validators.state.ValidatorRegisteredCopy;
+import com.radixdlt.atom.SubstateTypeId;
+import com.radixdlt.atom.TxBuilder;
+import com.radixdlt.constraintmachine.SubstateIndex;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.utils.KeyComparator;
+import com.radixdlt.utils.UInt256;
+import java.nio.ByteBuffer;
+import java.util.*;
+
+public record EpochConstructionState(
+    TxBuilder txBuilder,
+    TreeMap<ECPublicKey, ValidatorScratchPad> validatorsToUpdate,
+    RoundData closedRound,
+    EpochData closingEpoch,
+    TreeMap<ECPublicKey, TreeMap<REAddr, UInt256>> preparingStake,
+    NextEpochConstructorV3 constructor) {
+  public static EpochConstructionState createState(
+      NextEpochConstructorV3 constructor, TxBuilder txBuilder) {
+    return new EpochConstructionState(
+        txBuilder,
+        new TreeMap<>(KeyComparator.instance()),
+        txBuilder.downSystem(RoundData.class),
+        txBuilder.downSystem(EpochData.class),
+        new TreeMap<>(KeyComparator.instance()),
+        constructor);
+  }
+
+  public void upValidatorStakeData() {
+    validatorsToUpdate.values().forEach(scratchPad -> txBuilder.up(scratchPad.toSubstate()));
+  }
+
+  public void finalizeConstruction() {
+    txBuilder.up(new EpochData(closingEpoch.epoch() + 1));
+    txBuilder.up(new RoundData(0, closedRound.timestamp()));
+    txBuilder.end();
+  }
+
+  public void processExittingStake() {
+    var index = exittingStakeIndex(closingEpoch());
+
+    txBuilder
+        .shutdownAll(index, EpochConstructionState::collectExittingStake)
+        .forEach(stake -> txBuilder().up(stake.unlock()));
+  }
+
+  private ValidatorScratchPad stakeData(ECPublicKey publicKey) {
+    return validatorsToUpdate.computeIfAbsent(
+        publicKey, key -> new ValidatorScratchPad(txBuilder.down(ValidatorStakeData.class, key)));
+  }
+
+  private static SubstateIndex<ExitingStake> exittingStakeIndex(EpochData closingEpoch) {
+    var unlockedStateIndex =
+        ByteBuffer.allocate(2 + Long.BYTES)
```

Comment:
Can we have a variable for this and the other parameters with descriptive names?